### PR TITLE
Add support for building on ARM

### PR DESCRIPTION
**Pull request by ghost**
This patch was sourced from a FreeBSD PR

Source: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=201227
Author: Mikael Urankar

### CODE REVIEW

File: deps/v8/src/arm/assembler-arm.cc [hunk 1]
```diff
@@ -50,6 +50,35 @@ bool CpuFeatures::initialized_ = false;
 unsigned CpuFeatures::supported_ = 0;
 unsigned CpuFeatures::found_by_runtime_probing_ = 0;
 
+#ifdef __arm__
+
+bool OS::ArmCpuHasFeature(CpuFeature feature) {
```
Comment (misterdjules) [inline deps/v8/src/arm/assembler-arm.cc]: Why defining these functions in the ARM assembler for all OSes? Wouldn't that change break ARM support on platforms other than the one for which it was specifically written?

File: deps/v8/src/arm/assembler-arm.cc [hunk 2]
```diff
@@ -50,6 +50,35 @@ bool CpuFeatures::initialized_ = false;
...
+  return false;
+}
+
+CpuImplementer OS::GetCpuImplementer() {
+  static bool use_cached_value = false;
+  static CpuImplementer cached_value = UNKNOWN_IMPLEMENTER;
+  if (use_cached_value) {
+    return cached_value;
+  }
+  cached_value = ARM_IMPLEMENTER;
```
Comment (misterdjules) [inline deps/v8/src/arm/assembler-arm.cc]: Does FreeBSD support querying the implementer code instead of hardcoding it?

File: deps/v8/src/arm/assembler-arm.cc [hunk 3]
```diff
@@ -749,7 +778,7 @@ static bool fits_shifter(uint32_t imm32,
                          Instr* instr) {
   // imm32 must be unsigned.
   for (int rot = 0; rot < 16; rot++) {
-    uint32_t imm8 = (imm32 << 2*rot) | (imm32 >> (32 - 2*rot));
+    uint32_t imm8 = rot == 0 ? imm32 : ((imm32 << 2*rot) | (imm32 >> (32 - 2*rot)));
```
Comment (misterdjules) [inline deps/v8/src/arm/assembler-arm.cc]: Is this change really needed? It seems that both shift operations would zero-fill when applied to a `uint32_t`, and thus when `rot == 0`, this operation would be equivalent to `imm32 | 0`, or `imm32`, but I may be missing something.

File: deps/v8/src/arm/cpu-arm.cc [hunk 4]
```diff
@@ -64,7 +64,7 @@ void CPU::FlushICache(void* start, size_t size) {
   // None of this code ends up in the snapshot so there are no issues
   // around whether or not to generate the code when building snapshots.
   Simulator::FlushICache(Isolate::Current()->simulator_i_cache(), start, size);
-#else
+#elif defined(__linux__)
```
Comment (misterdjules) [inline deps/v8/src/arm/cpu-arm.cc]: Would using `#elif !defined(__FreeBSD__)` be more appropriate here? If I understand correctly, before this change this code is compiled for all platforms where `deps/v8/src/arm/cpu-arm.cc` is compiled, and thus not necessarily only on Linux.

File: deps/v8/src/atomicops.h [hunk 5]
```diff
@@ -160,8 +160,10 @@ Atomic64 Release_Load(volatile const Atomic64* ptr);
 #elif defined(__GNUC__) && \
   (defined(V8_HOST_ARCH_IA32) || defined(V8_HOST_ARCH_X64))
 #include "atomicops_internals_x86_gcc.h"
-#elif defined(__GNUC__) && defined(V8_HOST_ARCH_ARM)
+#elif defined(__GNUC__) && defined(__linux__) && defined(V8_HOST_ARCH_ARM)
```
Comment (misterdjules) [inline deps/v8/src/atomicops.h]: Same comment as [above](https://github.com/joyent/node/pull/25641/files#r34118403).

File: deps/v8/src/platform-freebsd.cc [hunk 6]
```diff
@@ -415,7 +415,7 @@ void* VirtualMemory::ReserveRegion(size_t size) {
   void* result = mmap(OS::GetRandomMmapAddr(),
                       size,
                       PROT_NONE,
-                      MAP_PRIVATE | MAP_ANON | MAP_NORESERVE,
+                      MAP_PRIVATE | MAP_ANON,
```
Comment (misterdjules) [inline deps/v8/src/platform-freebsd.cc]: For the record, this seems to be a backport of https://codereview.chromium.org/1025823003/, is that correct?

**Comment by misterdjules** [comment]
@feld Thank you for your interest in running node v0.10 on ARMv6 and FreeBSD. I added a few comments, mostly so that I can understand some of the details in this patch.

Let's see what this patch would look like at the end of the review process. At that time we'll be in a better place to evaluate whether it could land in this repository or should be kept as a patch in FreeBSD's ports.

File: deps/v8/src/arm/assembler-arm.cc (hunk 3 above)
Comment (MIkaelU) [inline deps/v8/src/arm/assembler-arm.cc]: This is an upstream patch (we are using clang on FreeBSD):
https://codereview.chromium.org/979633002/

File: deps/v8/src/platform-freebsd.cc (hunk 6 above)
Comment (MIkaelU) [inline deps/v8/src/platform-freebsd.cc]: Correct

File: deps/v8/src/arm/cpu-arm.cc (hunk 4 above)
Comment (MIkaelU) [inline deps/v8/src/arm/cpu-arm.cc]: From my understanding __ARM_NR_cacheflush is linux specific

File: deps/v8/src/atomicops.h (hunk 5 above)
Comment (MIkaelU) [inline deps/v8/src/atomicops.h]: code in atomicops_internals_arm_gcc.h is really linux/arm specific

File: deps/v8/src/arm/cpu-arm.cc (hunk 4 above)
Comment (misterdjules) [inline deps/v8/src/arm/cpu-arm.cc]: Ok, makes sense :+1:

File: deps/v8/src/atomicops.h (hunk 5 above)
Comment (misterdjules) [inline deps/v8/src/atomicops.h]: Right, sounds good :+1:

**Comment by misterdjules** [comment]
@MIkaelU Thank you very much for the feedback! :+1: Any comment on https://github.com/joyent/node/pull/25641#discussion_r34117960 and https://github.com/joyent/node/pull/25641#discussion_r34118231?

**Comment by feld** [comment]
good collaboration guys :+1:

File: deps/v8/src/arm/assembler-arm.cc (hunk 1 above)
Comment (MIkaelU) [inline deps/v8/src/arm/assembler-arm.cc]: I've screwed up this part of the patch... These changes should go in deps/v8/src/platform-freebsd.cc

File: deps/v8/src/arm/assembler-arm.cc (hunk 2 above)
Comment (MIkaelU) [inline deps/v8/src/arm/assembler-arm.cc]: No

**Comment by jasnell** [comment]
@misterdjules ... is this something we're going to land in v0.10?

**Comment by rvagg** [comment]
@nodejs/lts & @nodejs/build I'm inclined to say we shouldn't land this given that v4+ has proper ARM support and adding it back to v0.10 is just adding additional maintenance burden, thoughts?

**Comment by jasnell** [comment]
Agreed

**Comment by cjihrig** [comment]
I'd say this isn't appropriate for LTS or maintenance releases.

**Comment by rvagg** [comment]
done

sorry @feld, hopefully you're sorted on v4.x now though
